Make the EditorFileSystem cache more reliable

Rindbee opened the pull request:
Closes godotengine/godot-proposals#13196

## Known limitations

Due to the limited precision of file modification timestamps obtained by Godot, file movements are primarily tracked using UIDs. However, not all files are assigned a UID.

<details>

<summary>Details</summary>

### The file timestamps are not precise enough

File change detection relies on the file's modification timestamp. However, the timestamp precision of the current file's modification time is **1 second**.

And when processing files in batches, the modification timestamps of many files may all fall within the same second. This is especially noticeable when updating files using version control software such as **Git**.

This is unlikely to be a problem in most cases. However, some unexpected issues may arise in cases of file overwriting. For example:

1. If similar directory structures exist, such as `res://a/x/` and `res://b/x/`.
2. You have already added these changes using `git` and committed them.
3. You accidentally removed the directories `res://a/` and `res://b/`.
4. Using `git restore .` to restore these directories might be a good idea.
5. After restoration, the timestamps of these restored files and directories are typically within the same second.
6. Back to the editor. The timestamps of these files will be recorded.
7. If you delete `res://a/` outside the editor and rename the original `res://b/` to `res://a/`.
8. Back to the editor again. The content under `res://a/x/` is likely not updated correctly.

#### Solutions

1. Ensure that each resource file is assigned a unique UID;
2. Determine if a file has been updated by relying on its modification time and UID (if both of these remain unchanged, then the file is considered unchanged; otherwise, the file is considered to have changed);
3. Always force checks for changes in directories (i.e., the addition or deletion of sub files/directories) in projects managed with `git` (identify via `res://.git/`).

### Some types of file move actions cannot be tracked because no UID was assigned

The order in which file extensions are processed is as follows:

1. Importable file types (Their `importer` is either `skip` or `keep`);
2. Resource file type (The file extension list is `json,po,mo,crt,pub,key`, #99540);
3. Text file types (The file extension list is `txt,md,cfg,ini,log,json,yml,yaml,toml,xml`, defined in `docks/filesystem/textfile_extensions`);
4. Other file types (The file extension list is `ico,icns`, defined in `docks/filesystem/other_file_extensions`);

Point 2 and Point 3 are somewhat repetitive. If we process 3 and 4 first, point 2 might be able to be merged into point 3.

#### TODO

However, from the perspective of using UID to track file movements, it might be better to assign UIDs to files with the above extensions.

### The directory move actions could not be tracked because no UID was assigned

This may affect the updating of directory colors, favorites.

#### TODO

Perhaps the UID assigned to the directory could be recorded in a `.uid` file.

### The `@icon` path for the script

Using the `uid://` path allows for effective tracking; otherwise, it might be difficult to analyze the movement of `@icon` files. Furthermore, there is no method to update the `@icon` path in the script files.

</details>

## Main processing flow

### Main flowchart

<details>

<summary>Flowchart</summary>

```mermaid
flowchart TD
    S0(["Launch editor (first scan)"]) --> P00["Update global script classes"]
    P00 --> PA00["Scan"]
    PA00 --> C00{Already build tree?}
    C00 --> |No| PB03[[Build filesystem tree]]
    C00 --> |Yes| PB02[[Update filesystem tree]]
    PB02 --> PE00["Scan dirs"]
    PB03 --> PM01[Update uid actions]

    S1(["Application focus in"]) --> PB01["Scan full changes"]
    S2([Extensions changed]) --> PB01
    PB01 --> C01{Is scanning?}
    C01 --> |No|PB12["Mark fs dirty"]
    C01 --> |Yes|PB11["Full scan pending"]
    PB11 --> |Delayed to the end of the scan|PB01
    PB12 --> PD00["Scan dirty dirs"]

    S3([Dir operations in editor]) --> PB13["Mark dir dirty"]
    PB13 --> |Delayed to the end of the frame| PD00

    PD00 -->PE00
    PE00 -->PM01

    S4([File operations in editor]) --> PC00["Update file(s)"]
    PC00 --> |Delayed to the end of the frame|PM01

    PM01 --> C02{"Is first scan?"}
    C02 --> |No| PM02[Update global script classes]
    PM02 --> PM03[[Update scan actions]]
    C02 --> |Yes| PM03
    PM03 --> Z([End])
```

</details>

### Trigger a scan/update

Several cases that trigger a scan/update:

1. Launch editor (scan fs recursively);
2. Application focus in (scan fs recursively);
3. Extensions changed (scan fs recursively);
4. Dir operations (scan specific dirs);
5. File operations (update and/or scan specific files and/or dirs);
6. Some plugins actively call certain functions (update and/or scan).

<details>

<summary>Details</summary>

In a sense, Case 1 is a special case of Case 2. One difference is that the file information is stored in the file in case 1, while it is stored in memory in case 2. Another difference is that in case 1, the editor is being launched, so it may read the file information first, rather than reading the file information in the cache file to ensure that certain modules are available.

In case 3, the fact that the directory timestamp has not changed should not be an obstacle to detecting new files. The reasons for the changes in the list of supported file extensions are likely as follows:

1. Modified the editor settings `docks/filesystem/textfile_extensions` and/or `docks/filesystem/other_file_extensions`.
2. Modify certain scripts that inherit from `ResourceFormatLoader`[^1];
3. Activate/deactivate certain importer plugins[^2];

[^1]: These scripts must be global script classes (`class_name`) and must be tool (`@tool`) scripts.

[^2]: Custom importer scripts must inherit from `EditorImportPlugin` and be tool (`@tool`) scripts, and use `add_import_plugin()`/`remove_import_plugin()` to add/remove them.

Cases 4 and 5 are usually caused by user interaction in the editor. Some file caches (such as uids and global script classes) are project-wide, but some processing methods are implemented on a file-by-file basis (depending on `update_file()`). Additionally, the current file timestamp precision is **1 second**, making frequent calls to some methods unsuitable (#110048). These are the reasons for delaying processing until the end of the frame. It should be noted that if these file/dir operations affect the scripts mentioned in case 3, a new full scan may be triggered when the processing is completed.

Case 6 is probably the most unpredictable. Here are just some suggestions:

1. File write operations are best performed in the main thread;
2. Avoid frequent fs scanning.

</details>

### Update file info

During the scanning process, most of the information in `FileInfo` is updated **immediately**. This information is generally **unrelated** to file movement. However, information that might be affected by file movement is updated with a delay.

<details>

<summary>Details</summary>

The `FileInfo.status` records the file status at the corresponding path:

1. File changes;
2. Type changes;
3. File categories;
4. Has custom uid support or not;
5. Whether it is treated as a resource file;
6. Whether its type is a Script;
7. Whether its type is a PackedScene;
8. Whether it is a global class alternative;
9. Whether it is the **active** global class alternative.

The `FileInfo.class_name.icon_path` is used not only for scripts but also for other types of resources. `FileInfo.resource_script_class` is used for any resource that has script with a global class name. These two might become invalid when the script changes. However, the script will definitely be in `FileInfo.deps`. Modify `Object::get_property_list()` to ensure that the first `dep` found in `_find_resources()` of `ResourceFormatSaverBinaryInstance`/`ResourceFormatSaverTextInstance` is the `script`.

</details>

### Update UID and track paths

File changes can be tracked by monitoring changes in the UID. Since we obtain a snapshot of the file system during the scan, we need to analyze the changes in UIDs in a specific order to infer the changes to the files.

1. Verify whether the `UID(f)` is cached in `ResourceUID`;
2. Create and add a new UID;
3. Remove UID;
4. Attempting to add `UID(f)`;
5. If `UID(f)` already exists in `ResourceUID`, try adding `UID(m)`; if it still exists, try creating and adding a new UID.

<details>

<summary>Details</summary>

The above order is derived by working backward from the file changes. We may store the UID in both the `FileInfo` and the file itself. Read the new UID from the file to update the UID in memory.

Compound operations can be viewed as deleting first, and the UID is unique. Adding a UID **before** removing the existing one ensures the uniqueness of the newly added UID.

```mermaid
flowchart LR

A001(["Check uid change"]) --> C001{"Can have uid now?"}
C001 --> |Yes|A003["Check file status"]
C001 --> |No|A002["Remove UID(m) if valid"]
A002 --> E003([3])
A003 --> C002{"File removed?"}
C002 --> |Yes|A002
C002 --> |No|C003{"File added?"}
C003 --> |No|C005{"UID(m) valid?"}
C005 --> |Yes|C006{"UID(m) == UID(f)?"}
C006 --> |No|A008["Remove UID(m)"]
C006 --> |Yes|A007["Try to add UID(f) during first scan"]
A007 --> E001([1])
A008 --> E003
A008 --> C007{"UID(f) valid?"}
C005 --> |No|C007
C007 --> |Yes|A009["Try to add UID(f), UID(m)"]
A009 --> E004([4])
C007 --> |No|A010["Try to add UID(m)"]
A010 --> E004
C003 --> |Yes|C004{"UID(f) valid?"}
C004 --> |No|A006["Create a new UID"]
A006 --> E002([2])
C004 --> |Yes|A005["Try to add UID(f)"]
A005 --> E004
```

</details>

### Analyze file changes based on path changes in UID

During each scan, we record the basic File operations in `FileInfo.status` and use UID to track path changes.

Based on the above UID tracking records, the file operations are inferred according to rules.

<details>

<summary>Details</summary>

File operation types include (`U` for UID, `P` for path):

1. Basic operations:
   1. File added (UID added), `PaUa`;
   2. File removed (UID removed), `PrUr`;
   3. File updated, `Pu`;
      1. UID added, `Ua`;
      2. UID removed, `Ur`;
      3. UID unchanged.
2. Compound operations:
   1. File moved/renamed, `P0U0r + P1U0a`;
   2. File overwriten, `P0U0r + P1U1r + P1U0a + ...`;
   3. File duplicated, `P0U0r + P0U0a + P1U0a + ...`.

</details>

### Update the caches after confirming the file has been moved

1. Update resource paths of instances in `ResourceCache`;
2. Update global script classes;
3. Update deps, resource script class, and icons;
4. Update project settings.

## Jobs

### Multi-dot extensions

Improve support for multi-dot extensions.

<details>

<summary>Tasks</summary>

- [x] Add a `String::validate_extension()` method to validate whether the path's extension is in the provided list of extensions.
- [x] Improved support for multi-dot extensions in the file system.
- [x] Improved support for multi-dot extensions in file dialogs.

</details>

### File status

Use `status` to record information such as the file category, type, and status.

<details>

<summary>Tasks</summary>

- [x] Classification.
  - [x] Custom uid support, tagged in `EditorFileSystem::_category_validate()`.
  - [x] Categorize files by file extension, tagged in `EditorFileSystem::_category_validate()`.
    - [x] Importable files.
    - [x] Text files.
    - [x] Other files.
  - [x] Classified by `type`, tagged in `EditorFileSystem::_type_analysis()`.
    - [x] As resources.
    - [x] Not a resource.
      - [x] OtherFile.
      - [x] TextFile.
      - [x] Empty type.
- [x] The existence status of files on the same path changes.
  - [x] File removed, tagged in `EditorFileSystem::_file_info_remove()`.
  - [x] File added, handled in `EditorFileSystem::_file_info_add()`.
  - [x] File updated, handled in `EditorFileSystem::_file_info_update()`.
- [x] Type changes, tagged in `EditorFileSystem::_type_analysis()`.
  - [x] Type removed.
    - [x] Clear the path of the cached instance, handled in `EditorFileSystem::_update_scan_actions()`.
  - [x] Type added.
  - [x] Type changed, equivalent to removing first and then adding.
    - [x] Clear the path of the cached instance, handled in `EditorFileSystem::_update_scan_actions()`.
  - [x] Type unchanged, just a file update.
    - [x] Reload the cached instance.
      - [x] Scripts.
      - [x] Scenes.
      - [x] Other resources, sent via the `resources_reload` signal, handled in `EditorNode::_resources_changed()`.

</details>

### File movement

Track changes to UIDs to analyze complex operations such as file movement.

<details>

<summary>Tasks</summary>

- [x] Sort the processing order based on changes in UID, handled in `EditorFileSystem::_create_actions_from_uid_change()`.
- [x] Update `ResourceUID` cache in order, handled in `EditorFileSystem::_update_scan_uid_actions()`.
- [x] Track file changes based on changes in UID, handled in `EditorFileSystem::_update_scan_uid_actions()`.
  - [x] Categorizing file changes.
    - [x] File untracked, the UID is no longer in use.
    - [x] File moved, the file path using the UID has changed.
    - [x] File duplicated, a new file path attempting to use the existing UID has been detected.
    - [x] File tracked, the UID was used for the first time.
    - [x] File retracked, the UID is already included in the file.
  - [ ] Process caches based on UID changes.
    - [x] Resource path in cached instances.
      - [x] Untracked files.
        - [x] Cleared.
      - [x] Moved files.
        - [x] Updated.
          - [x] Use the `.tmp~` temporary suffix to avoid path swapping cases.
    - [x] File owners and dependencies.
      - [x] Untracked files.
        - [x] Keep.
      - [x] Moved files.
        - [x] Update the `deps` in the owner's file info.
        - [x] Rename dependencies in the owner files.
        - [x] When the affected file owner is an opened scene file, reload occurs.
    - [x] Project settings.
      - [x] Update affected file paths.
    - [ ] Folder tracking has not yet been implemented. Use a `.uid` file to track?
      - [ ] Folder colors.
        - [x] In editor, call `FileSystemDock::_update_folder_colors_after_move()`.
        - [ ] Outside editor.
      - [ ] Favorites.
    - [ ] Update `ResourceCache::resource_path_cache`.
    - [ ] Provide a dialog that displays the changes.

</details>

### Script class info

Update class information in the script.

<details>

<summary>Tasks</summary>

- [x] Manage class info, handled in `EditorFileSystem::_script_class_info_update()`.
  - [x] Docs.
  - [x] Icons.
    - [x] Scripts.
      - [x] `class_info.icon_path`[^3].
    - [x] Resource files.
      - [x] `class_info.icon_path`[^3].
- [x] Manage global script class.
  - [x] Cache global script class alternatives.
  - [x] Handles updates for global script alternatives.
    - [x] Class info removed, handled in `EditorFileSystem::_global_script_class_info_remove()`.
    - [x] Class info added, handled in `EditorFileSystem::_global_script_class_info_add()`.
    - [x] Class info updated, equivalent to removing first and then adding, handled in `EditorFileSystem::_script_class_info_update()`.
  - [x] Maintaining active global script classes.
    - [x] Activate, handled in `EditorFileSystem::_update_global_script_class_activation()`.
      - [x] Register in `ScriptServer`.
      - [x] Register in `EditorData`.
        - [x] Class name.
        - [x] Icon.
      - [x] Generate its documentation.
    - [x] Deactivate, handled in `EditorFileSystem::_global_script_class_info_remove()`.
      - [x] Unregister in `ScriptServer`.
      - [x] Unregister in `EditorData`.
        - [x] Class name.
        - [x] Icon.
      - [x] Clear its documentation.
    - [x] Special class derived scripts.
      - [x] `ResourceFormatLoader`.
      - [x] `ResourceFormatSaver`.

[^3]: The icon displayed as the file icon conflicts with the icon specified in the script's icon field.

</details>

### Scenes

Update global scene groups and built-in scripts.

<details>

<summary>Tasks</summary>

- [x] Global scene groups.
- [x] Built-in scripts.
  - [x] Docs.

</details>

### Importable files

<details>

<summary>Tasks</summary>

- [x] Verify whether (re)importing is necessary, handled in `EditorFileSystem::_import_validate()`.

</details>

## Known issues

<details>

<summary>An error occurs when updating the tscn file that uses ViewportTexture</summary>

```bash
ERROR: Path to node is invalid: 'SubViewportContainer/SubViewport'.
   at: _setup_local_to_scene (scene/main/viewport.cpp:200)
```

`EditorFileSystem::_update_script_documentation()` will load and use the `PackedScene`. Resources that are not configured with a local scene in the editor will return the root of the currently edited scene when `Resource::get_local_scene()` is called.

</details>

Rindbee committed:
Add UID suport in global script class cache

Unify `global_script_class_cache.cfg` saving logic

This is the first step to making `global_script_class_cache.cfg` reliable.

Use uid instead of path in the global script class cache

AThousandShips commented on editor/file_system/editor_file_system.h:
```diff
@@ -149,51 +181,99 @@ class EditorFileSystem : public Node {
+	void _script_class_info_update(EditorFileInfo *p_file, const String &p_path, const ScriptClassInfo *p_sci);
+
+	void _file_info_add(EditorFileSystemDirectory *p_parent_dir, const String &p_parent_path, const String &p_file, bool p_insert);
+	void _file_info_remove(EditorFileInfo *p_file, const String &p_path, const int p_idx);
```
```suggestion
	void _file_info_remove(EditorFileInfo *p_file, const String &p_path, int p_idx);
```

AThousandShips commented on editor/file_system/editor_file_system.h:
```diff
@@ -149,51 +181,99 @@ class EditorFileSystem : public Node {
+	void _file_info_remove(EditorFileInfo *p_file, const String &p_path, const int p_idx);
+	void _file_info_update(EditorFileInfo *p_file, const String &p_path);
+
+	int _dir_info_remove(EditorFileSystemDirectory *p_dir, const String &p_path, const int p_idx);
```
```suggestion
	int _dir_info_remove(EditorFileSystemDirectory *p_dir, const String &p_path, int p_idx);
```

AThousandShips commented on core/io/resource_uid.cpp:
```diff
@@ -281,8 +296,12 @@ Error ResourceUID::load_from_cache(bool p_reset) {
 		c.saved_to_cache = true;
 		unique_ids[id] = c;
+
+		TOOLS_PRINT_VERBOSE(vformat("Loading uid:\"%s\" path:\"%s\".", id_to_text(id), String::utf8(c.cs.ptr())));
```
```suggestion
		TOOLS_PRINT_VERBOSE(vformat(R"(Loading UID: "%s" path: "%s".)", id_to_text(id), String::utf8(c.cs.ptr())));
```

AThousandShips commented on core/io/resource_uid.cpp:
```diff
@@ -314,9 +337,10 @@ Error ResourceUID::update_cache() {
 		}
 	}
-
+	TOOLS_PRINT_VERBOSE("Appending uids done.");
```
```suggestion
	TOOLS_PRINT_VERBOSE("Appending UIDs done.");
```

AThousandShips commented on core/io/resource_uid.cpp:
```diff
@@ -314,9 +337,10 @@ Error ResourceUID::update_cache() {
 			f->store_buffer((const uint8_t *)E.value.cs.ptr(), s);
 			E.value.saved_to_cache = true;
 			cache_entries++;
+			TOOLS_PRINT_VERBOSE(vformat("Appending uid:\"%s\" path:\"%s\".", id_to_text(E.key), String::utf8(E.value.cs.ptr())));
```
```suggestion
			TOOLS_PRINT_VERBOSE(vformat(R"(Appending UID: "%s" path: "%s".)", id_to_text(E.key), String::utf8(E.value.cs.ptr())));
```

AThousandShips commented on core/io/resource_uid.cpp:
```diff
@@ -268,6 +281,8 @@ Error ResourceUID::load_from_cache(bool p_reset) {
 		unique_ids.clear();
 	}
 
+	TOOLS_PRINT_VERBOSE("Loading uids from cache...");
```
```suggestion
	TOOLS_PRINT_VERBOSE("Loading UIDs from cache...");
```

AThousandShips commented on core/io/resource_uid.cpp:
```diff
@@ -244,15 +252,20 @@ Error ResourceUID::save_to_cache() {
+		TOOLS_PRINT_VERBOSE(vformat("Saving uid:\"%s\" path:\"%s\".", id_to_text(E.key), String::utf8(E.value.cs.ptr())));
 	}
 
+	TOOLS_PRINT_VERBOSE("Saving uids done.");
```
```suggestion
	TOOLS_PRINT_VERBOSE("Saving UIDs done.");
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -840,7 +931,215 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+	remove_point = scan_actions.push_back(ia);
+}
+
+void EditorFileSystem::_create_action(EditorFileSystemDirectory *p_dir, EditorFileInfo *p_fi, const String &p_path, const ItemAction::Action p_action, const ItemAction::Step p_step, const ResourceUID::ID p_old_uid) {
```
```suggestion
void EditorFileSystem::_create_action(EditorFileSystemDirectory *p_dir, EditorFileInfo *p_fi, const String &p_path, ItemAction::Action p_action, ItemAction::Step p_step, ResourceUID::ID p_old_uid) {
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -840,7 +931,215 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+				continue;
+			} break;
+			case ItemUIDAction::ACTION_UID_ADD: {
+				print_verbose(vformat("[%.6f] [ADD UID] old uid: %s, uid: %s, path: %s.",
```
```suggestion
				print_verbose(vformat(R"([%.6f] [ADD UID] old UID: "%s", UID: "%s", path: "%s".)",
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -840,7 +931,215 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+				print_verbose(vformat("[ADD UID] %s, %s", ia.path, ResourceUID::get_singleton()->id_to_text(ia.file->uid)));
+			} break;
+			case ItemUIDAction::ACTION_UID_REMOVE: {
+				print_verbose(vformat("[%.6f] [REMOVE UID] old uid: %s, uid: %s, path: %s.",
```
```suggestion
				print_verbose(vformat(R"([%.6f] [REMOVE UID] old UID: "%s", UID: "%s", path: "%s".)",
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -1194,238 +1407,417 @@ int EditorFileSystem::_scan_new_dir(ScannedDirectory *p_dir, Ref<DirAccess> &da)
+	ERR_FAIL_NULL(p_file);
+	ScriptClassInfo &sci = p_file->class_info;
+	HashMap<StringName, ScriptClassAlternatives>::Iterator E = global_script_class_alternatives.find(sci.name);
+	ERR_FAIL_NULL_MSG(E, vformat("The file %s of the untracked global class %s was detected and removed.", p_path, sci.name));
```
```suggestion
	ERR_FAIL_NULL_MSG(E, vformat(R"(The file "%s" of the untracked global class "%s" was detected and removed.)", p_path, sci.name));
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -1450,83 +1842,65 @@ void EditorFileSystem::_scan_fs_changes(EditorFileSystemDirectory *p_dir, ScanPr
+					}
+
+					if (first_scan) {
+						Vector<String> dirs = dir_path.substr(6, dir_path.length() - 6).split("/");
```
```suggestion
						Vector<String> dirs = dir_path.substr(6).split("/");
```
Redundant, that's the full remaining string

AThousandShips commented on editor/file_system/editor_file_system.h:
```diff
@@ -268,18 +329,28 @@ class EditorFileSystem : public Node {
+	List<ItemUIDAction>::Element *uid_move_end;
 	List<ItemAction> scan_actions;
+	List<ItemAction>::Element *normal;
+	List<ItemAction>::Element *remove_point;
```
```suggestion
	List<ItemAction>::Element *normal = nullptr;
	List<ItemAction>::Element *remove_point = nullptr;
```

AThousandShips commented on doc/classes/EditorFileSystem.xml:
```diff
@@ -54,21 +54,20 @@
 			<return type="void" />
 			<description>
-				Scan the filesystem for changes.
+				Scan the filesystem for changes. During the scan, the timestamps of the folders will be ignored, and the scan will always try to detect new changes to the folders.
```
```suggestion
				Scans the filesystem for changes. During the scan, the timestamps of the folders will be ignored, and the scan will always try to detect new changes to the folders.
```

AThousandShips commented on doc/classes/EditorFileSystem.xml:
```diff
@@ -54,21 +54,20 @@
 			<return type="void" />
 			<description>
-				Check if the source of any imported resource changed.
+				Scan the filesystem for changes. The behavior is affected by [member EditorSettings.filesystem/on_scan/detect_mode].
```
```suggestion
				Scans the filesystem for changes. The behavior is affected by [member EditorSettings.filesystem/on_scan/detect_mode].
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -1194,238 +1463,416 @@ int EditorFileSystem::_scan_new_dir(ScannedDirectory *p_dir, Ref<DirAccess> &da)
+		return;
+	}
+	// Must not reimport, all was good.
+	// Update modified times, md5 and destination paths, to avoid reimport.
```
```suggestion
	// Update modified times, md5, and destination paths, to avoid reimport.
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -1019,36 +1285,42 @@ bool EditorFileSystem::_update_scan_actions() {
-
 	if (first_scan) {
-		//only on first scan this is valid and updated, then settings changed.
+		// Only on first scan this is valid and updated, then settings changed.
```
```suggestion
		// Only on the first scan is this valid and updated, then settings changed.
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -840,7 +933,273 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+		};
+		HashMap<ResourceUID::ID, MoveItems> moves;
+
+		print_verbose("======= Scan UID Analysis Start ======");
```
```suggestion
		print_verbose("======= Scan UID Analysis Start =======");
```

Rindbee committed:
Disconnect signals in `GDScriptLanguage::finish()`

- extension_loaded
- extension_unloading

These signals are best disconnected in `GDScriptLanguage::finish()`
to prevent spam in unit tests.

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -1019,36 +1289,42 @@ bool EditorFileSystem::_update_scan_actions() {
 		reimport_files(reimports);
 	} else {
-		//reimport files will update the uid cache file so if nothing was reimported, update it manually
+		// Reimport files will update the uid cache file so if nothing was reimported, update it manually.
```
```suggestion
		// Reimporting files will update the UID cache file so if nothing was reimported, update it manually.
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -840,166 +938,487 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+					// In legacy projects, the UID might be missing. But it shouldn't be repaired here.
+					// TODO: An interactive repair dialog should be provided.
+					new_deps.push_back(dep);
+					ERR_PRINT_ONCE(vformat("The uid is missing from the path %s in the deps of %s.", dep, efsd->get_path().path_join(fi->file)));
```
```suggestion
					ERR_PRINT_ONCE(vformat("The UID is missing from the path %s in the deps of %s.", dep, efsd->get_path().path_join(fi->file)));
```

Rindbee committed:
Make the EditorFileSystem cache more reliable

Rindbee committed:
Improve multi-dot extensions support

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -840,166 +933,527 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+					is_first_dep = false;
+					int idx = -1;
+					EditorFileSystemDirectory *dep_dir = find_file(dep_path, &idx);
+					ERR_CONTINUE_MSG(!dep_dir, vformat("The dir info of %s was not found when checking %s", dep_path, file_path));
```
```suggestion
					ERR_CONTINUE_MSG(!dep_dir, vformat("The dir info of %s was not found when checking %s.", dep_path, file_path));
```

Rindbee committed:
Allow updating the `script_class` field

Add `ResourceFormatSaver::set_script_class()` to allow updating the
`script_class` field.

Supported file types:

- tres (`ResourceFormatSaverText`)
- res (`ResourceFormatSaverBinary`)

Rindbee committed:
The `_find_resources()` method prioritizes searching for `script` property.

Ensure that the `script` is marked as `1` in `_find_resources()`.

This ensures that the first script resource in `deps` is `script`
when only parsing is performed and not loaded.

TODO: Properly clean up `ResourceCache::resource_path_cache` to
prevent misjudgments when changing scripts.

Rindbee committed:
Provide aliases for commonly used structure names in `EditorFileSystem`

- `EditorFileInfo` for `EditorFileSystemDirectory::FileInfo`
- `ScriptClassInfo` for `EditorFileSystemDirectory::FileInfo::ScriptClassInfo`

Rindbee committed:
Revert "Core: Do not generate `*.uid` files for JSON, certificates, and translations"

This reverts commit b91bacb43af4f797981c7933f944da2ad6c0c06d.

AThousandShips commented on core/io/resource_uid.cpp:
```diff
@@ -244,15 +252,20 @@ Error ResourceUID::save_to_cache() {
 
 	cache_entries = 0;
 
+	TOOLS_PRINT_VERBOSE("Saving uids to file...");
```
```suggestion
	TOOLS_PRINT_VERBOSE("Saving UIDs to file...");
```

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -840,7 +931,215 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+
+	EditorProgress *ep = nullptr;
+	if (scan_uid_actions.size() > (EditorFileSystem::ItemUIDAction::STEP_UID_MAX / 2 + 1)) {
+		ep = memnew(EditorProgress("_update_scan_uid_actions", TTR("Scanning uid actions..."), scan_uid_actions.size()));
```
```suggestion
		ep = memnew(EditorProgress("_update_scan_uid_actions", TTR("Scanning UID actions..."), scan_uid_actions.size()));
```

AThousandShips commented on core/io/resource_saver.cpp:
```diff
@@ -169,6 +167,21 @@ Error ResourceSaver::set_uid(const String &p_path, ResourceUID::ID p_uid) {
 }
 
+Error ResourceSaver::set_script_class(const String &p_path, const String &p_script_class) {
+	ERR_FAIL_COND_V_MSG(p_path.is_empty(), ERR_INVALID_PARAMETER, "Can't update UID to empty path. Provide non-empty path.");
```
```suggestion
	ERR_FAIL_COND_V_MSG(p_path.is_empty(), ERR_INVALID_PARAMETER, "Can't update the script class for an empty path. Provide a non-empty path.");
```

KoBeWi commented on core/io/json.cpp:
This seems to make .json files not editable from the inspector. Was it causing problems?

Rindbee replied:
Related: https://github.com/godotengine/godot/issues/105160#issuecomment-3759575237

However, it probably shouldn't be included in this PR.

KoBeWi commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -855,166 +939,531 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+					is_first_dep = false;
+					int idx = -1;
+					EditorFileSystemDirectory *dep_dir = find_file(dep_path, &idx);
+					ERR_CONTINUE_MSG(!dep_dir, vformat("The dependency path %s in file %s does not exist. Please edit its dependencies to fix the dependency path manually.", dep_path, file_path));
```
I'm getting this error for outdated path, where the UID is valid. The error should tell you to re-save resource instead of fixing the path manually (it's easy to misinterpret).

Also why is it getting printed on every editor focus? The file was not modified, so it should either happen only once, or only at launch, to avoid spamming.

Rindbee replied:
> I'm getting this error for outdated path, where the UID is valid. The error should tell you to re-save resource instead of fixing the path manually (it's easy to misinterpret).

Do the UIDs in the import/uid file match the UID records in the files that use them? The UID may have been changed in the past. If the UID is reliable, the path should be updated automatically.

> Also why is it getting printed on every editor focus? The file was not modified, so it should either happen only once, or only at launch, to avoid spamming. Or you could just force re-save of the parent file?

Is it only encountered when launching the editor? It's probably more likely to occur when using `git checkout` in older projects.

KoBeWi commented on editor/docks/filesystem_dock.cpp:
```diff
@@ -1906,9 +1898,10 @@ void FileSystemDock::_rename_operation_confirm() {
-	_update_project_settings_after_move(file_renames, folder_renames);
+	// _update_resource_paths_after_move(file_renames, uids);
+	// _update_dependencies_after_move(file_renames, file_owners);
+	// _update_project_settings_after_move(file_renames);
```
Why is this not removed completely?

KoBeWi commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -1209,238 +1659,431 @@ int EditorFileSystem::_scan_new_dir(ScannedDirectory *p_dir, Ref<DirAccess> &da)
+	}
+
+	if (p_file->status & EditorFileInfo::IS_IMPORTABLE) {
+		// TODO: Clean up the previous state.
```
This blocking?

KoBeWi commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -3783,21 +4220,37 @@ EditorFileSystem::EditorFileSystem() {
+	if (!force_detect) {
+		// This should probably also work on Unix and use the string it returns for FAT32 or exFAT
+		Ref<DirAccess> da = DirAccess::create(DirAccess::ACCESS_RESOURCES);
+		force_detect = da->get_filesystem_type() == "FAT32" || da->get_filesystem_type() == "EXFAT" || DirAccess::exists("res://.git");
```
What about other version control systems? Kinda weird to hard-code git.

KoBeWi reviewed (commented):
Looks like this PR is re-adding .uid files for some Resources that had them intentionally removed. This is questionable (although at least in case of translations I can agree).

There are some core changes that someone needs to review too. I noticed this PR also includes #110779, I guess that PR could be merged separately, if UIDs belong in ScriptServer. And #114944 too, but the code is a bit different? 🤔

As for editor changes... There is lots of them, so it's difficult to do a detailed review. I checked over all of the code and it seems fine, but also feels a bit over-engineered. Like, there are many new structs, enums, or fields in EditorFileSystem. It's difficult to test if this new code is more reliable, or if it introduces some edge-case bugs (like the error spam I discovered, which wasn't there before). It does look more robust at least.
I did see some random errors in the current editor file system, especially when moving scripts, so I guess this could use some broader testing to see if it resolves these bugs, while not breaking anything.

Rindbee commented on doc/classes/EditorFileSystem.xml:
```diff
@@ -60,21 +60,20 @@
 			<description>
-				Add a file in an existing directory, or schedule file information to be updated on editor restart. Can be used to update text files saved by an external program.
-				This will not import the file. To reimport, call [method reimport_files] or [method scan] methods.
+				Immediately update the file information from the file to the memory cache, which may trigger further actions at the end of the frame.
```
This method also applies to importable files. Primarily used for updating FileInfo. However, the import step is postponed to the end of the frame.

Rindbee commented on core/io/resource_loader.h:
```diff
@@ -251,7 +251,7 @@ class ResourceLoader {
 	static String get_resource_type(const String &p_path);
 	static String get_resource_script_class(const String &p_path);
 	static ResourceUID::ID get_resource_uid(const String &p_path);
-	static bool should_create_uid_file(const String &p_path);
```
Yes, it's used for categorizing by path. It doesn't check if the uid file exists.

Rindbee committed:
`get_modified_time()` returns the maximum value of `ctime` and `mtime` on Unix

Since `mv` and `cp -p` may preserve the `mtime`. And the OS file manager
seems to behave in accordance with these behaviors.

Files extracted from the same compressed file may have the same timestamp.

Therefore, when using these files to overwrite, from the perspective
of the path, the `mtime` may not change, even though the content may
have changed.

Rindbee committed:
Improved cache files cleanup when removing files

Imitater967 commented:
looking forward this merge, it stops me from using godot

AThousandShips commented on editor/file_system/editor_file_system.cpp:
```diff
@@ -857,177 +950,557 @@ bool EditorFileSystem::_scan_import_support(const Vector<String> &reimports) {
+							if (ResourceUID::get_singleton()->has_id(uid)) {
+								const String uid_path = ResourceUID::get_singleton()->get_id_path(uid);
+								if (uid_path != dep_path) {
+									WARN_PRINT(vformat("The path %s of dependency (UID %s) in file %s doesn's match the path %s in UID cache. Please fix this dependency promptly.", dep_path, uid_text, file_path, uid_path));
```
```suggestion
									WARN_PRINT(vformat("The path %s of dependency (UID %s) in file %s doesn't match the path %s in UID cache. Please fix this dependency promptly.", dep_path, uid_text, file_path, uid_path));
```

Rindbee committed:
Fix typo

Imitater967 commented:
still watching is pr
